contrib/submit-simple-job/App/convert.ts

```diff
@@ -0,0 +1,148 @@
+    if (key === "uid") { return uid; }
+    if (key === "gid") { return gid; }
+    throw Error("Bad replacement");
+  }).split("\n").join(" && ");
+
+  const userCommands = [];
+  if (simpleJob.root) {
+    userCommands.push(command);
+  } else {
+    // Create a user with group.
+    userCommands.push(
+      `groupadd --gid ${gid} ${user}`,
+      `useradd --gid ${gid} --groups sudo --uid ${uid} ${user}`,
+      "echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers",
+      `echo '${command.replace(/'/g, "'\\''")}' | su --preserve-environment ${user}`,
+    );
+    depends.sudo = true;
+  }
+
+  const dependencies = Object.keys(depends);
+  const installCommands = [];
+  if (dependencies.length > 0) {
+    installCommands.push(
+      "apt-get update",
+      `apt-get install --assume-yes ${dependencies.join(" ")}`,
+    );
+  }
+
+  const commands = installCommands
+    .concat(mountCommands)
+    .concat(exportCommands)
+    .concat(userCommands);
+
+  return commands.join(" && ");
+}
+
+function convertTaskRole(name: string, simpleJob: SimpleJob, user: string, hyperParameterValue?: number) {
+  const taskRole: any = {
+    command: convertCommand(simpleJob, user, hyperParameterValue),
+    gpuNumber: simpleJob.gpus,
+    name,
+  };
+
+  if (simpleJob.isPrivileged) {
+    taskRole.cpuNumber = simpleJob.cpus;
+    taskRole.memoryMB = simpleJob.memory;
+  }
+
+  if (simpleJob.isInteractive) {
+    const portList = [];
+    const ports = simpleJob.interactivePorts.split(/[;,]/);
+    for (const portString of ports) {
+      const port = Number(portString);
+      if (isNaN(port)) { continue; }
+      portList.push({
+        beginAt: port,
+        label: `port_${port}`,
+        portNumber: 1,
+      });
+    }
+    taskRole.portList = portList;
+  }
+
+  return taskRole;
+}
+
+export default function convert(simpleJob: SimpleJob, user: string) {
+  const job: any = {
+    image: simpleJob.image,
+    jobName: simpleJob.name,
+    virtualCluster: simpleJob.virtualCluster,
+  };
+
+  if (authFile !== undefined) {
+    job.authFile = authFile;
+  }
+
+  const taskRoles = [];
+
+  if (simpleJob.hyperParameterName === "") {
+    taskRoles.push(convertTaskRole("master", simpleJob, user));
+  } else {
+    for (
+      let hyperParameterValue = simpleJob.hyperParameterStartValue;
+      hyperParameterValue < simpleJob.hyperParameterEndValue;
+      hyperParameterValue += simpleJob.hyperParameterStep
+    ) {
+      const taskName = `hyper_parameter_${hyperParameterValue}`;
+      const taskRole = convertTaskRole(taskName, simpleJob, user, hyperParameterValue);
+      taskRoles.push(taskRole);
+    }
+  }
+
+  if (simpleJob.enableTensorboard) {
+    const tensorboardSimpleJob = new SimpleJob(simpleJob);
+    tensorboardSimpleJob.gpus = 0;
+    tensorboardSimpleJob.command = `tensorboard --logdir ${simpleJob.tensorboardModelPath} --host 0.0.0.0`;
+    tensorboardSimpleJob.isInteractive = true;
+    tensorboardSimpleJob.interactivePorts = "6006";
```

Comment:
There may be port conflict if hard code a port number here. Maybe add a new port in `portList`?

Reply:
In DLWorkspace's job design, all ports are hard-coded, since they use virtual network and do not have the port confliction issue.

Per our discussion, YARN should also schedule job with hard coded ports to nodes without same open ports.